### ISSUE TITLE
ODB: The OneDrive API is inconsistent with providing hash data in /delta JSON response (URGENT BUG)

### ISSUE THREAD

**Issue by abraunegg** [OneDrive/onedrive-api-docs#1704]
#### Category
- [ ] Question
- [ ] Documentation issue
- [X] Bug

The OneDrive API 

#### Expected or Desired Behavior

The `hashes` component should always have an entry.
```
"file": {
				"hashes": {
					"quickXorHash": "qlLLA1L3gC6HiNFu2JkgxT8ICuk="
				},
				"mimeType": "text/plain"
			},
```

#### Observed Behavior

As per https://github.com/abraunegg/onedrive/issues/2423 - the OneDrive Client for Linux now is only looking for `quickXorHash` or `sha256Hash` for files - however it appears in some instances the API does not always provide `hashes` in the /delta API response:
```
"file": {
				"hashes": {},
				"mimeType": "text/plain"
			},
```

![image](https://github.com/OneDrive/onedrive-api-docs/assets/4956234/4c3c5ede-79c5-4672-a3f9-af470977e87a)

#### Steps to Reproduce

Use the /delta query or the OneDrive Client for Linux  - see https://github.com/abraunegg/onedrive/issues/2433
Thank you.

[ ]: http://aka.ms/onedrive-api-issues
[x]: http://aka.ms/onedrive-api-issues

**Comment by ificator** [comment]
This looks to be an edge case with 0-byte files not generating a `quickXorHash` value.

**Comment by ificator** [comment]
A fix has been made and will deploy soon.

**Comment by Rmano** [comment]
@ificator thanks! Notice that I have errors of "no hash" also for file that are not 0-sized (they seems to come from One Note). If you need/want more info, I can try to dig it.

**Comment by ificator** [comment]
#1706 appears to be OneDrive Personal.

@abraunegg does your change fall back to sha1 if quickXor is missing?

**Comment by abraunegg** [comment]
> #1706 appears to be OneDrive Personal.
> 
> @abraunegg does your change fall back to sha1 if quickXor is missing?

No it does not - as it has been removed from the codebase and being stored.

The ms docs even talk about crc33 which was removed a long time ago.

**Comment by abraunegg** [comment]
@ificator 
The code looks at quickxorhash then sha256hash as the fallback

**Comment by pablobass** [comment]
> #1706 appears to be OneDrive Personal.

It is.

**Comment by ificator** [comment]
> > #1706 appears to be OneDrive Personal.
> 
> It is.
> 

@pablobass are you able to see a pattern with the files? File type, location, etc?

**Comment by Rmano** [comment]
Mine is a business account, but it *seems* to me (I have just the id, and the name of the directories) that my non-zero files causing the error are related to OneNote.

**Comment by pablobass** [comment]
> > > #1706 appears to be OneDrive Personal.
> > 
> > 
> > It is.
> 
> @pablobass are you able to see a pattern with the files? File type, location, etc?

I don't know how to identify wich files are referring the "hash" notices to.
I tried dual verbose option and it does not give me that information too.

How can I do to see it?

**Comment by abraunegg** [comment]
> > > > #1706 appears to be OneDrive Personal.
> > > 
> > > 
> > > It is.
> > 
> > 
> > @pablobass are you able to see a pattern with the files? File type, location, etc?
> 
> I don't know how to identify wich files are referring the "hash" notices to. I tried dual verbose option and it does not give me that information either.
> 
> What can I do to see it?

If you are seeing an 'id' rather than a 'path' this means the item in question is being excluded by your configuration.

**Comment by Rmano** [comment]
> If you are seeing an 'id' rather than a 'path' this means the item in question is being excluded by your configuration.

When I was playing with the code I think to remember that sometimes, even if the `->parent` field was not available, the `name` field could have some information... but I was really printing random things  so I can't guarantee it.

@pablobass in the "double verbose" log you should see something like this:

```
[DEBUG] Not a OneDrive deleted item change
[DEBUG] defaultRootId                                        = 01CRET2R56Y2GOVW7725BZO354PWSELRRZ
[DEBUG] 'search id'                                          = 01CRET2R56Y2GOVW7725BZO354PWSELRRZ
[DEBUG] id == defaultRootId                                  = true
[DEBUG] isItemRoot(item)                                     = false
[DEBUG] item['name'].str == 'root'                           = false
[DEBUG] singleDirectoryScope                                 = false
[DEBUG] 'parentReference id'                                 = 01CRET2R63B3UZEGQYDFFIZCQVNQAURE4X
[DEBUG] search criteria: syncFolderName                      = root
[DEBUG] search criteria: syncFolderPath                      = 
[DEBUG] search criteria: syncFolderChildPath                 = 
[DEBUG] thisItemId                                           = 01CRET2R2NSM7N4LKUYVGKU7PU3ZBUEZ73
[DEBUG] thisItemParentPath                                   = /drives/b!LtHke7bh3kG0bVzBE1QCWW33fbzkHnhNqI5o-VcAsHJ_LJBMtZ3mQagpnag-F9Cq/root:/Blocs de notas/Cuaderno de Moodle
[DEBUG] thisItemFullPath                                     = /drives/b!LtHke7bh3kG0bVzBE1QCWW33fbzkHnhNqI5o-VcAsHJ_LJBMtZ3mQagpnag-F9Cq/root:/Blocs de notas/Cuaderno de Moodle/Cuaderno de Moodle.onetoc2
[DEBUG] 'item id' matches search 'id'                        = false
[DEBUG] 'parentReference id' matches search 'id'             = false
[DEBUG] 'thisItemParentPath' contains 'syncFolderChildPath'  = true
[DEBUG] 'thisItemParentPath' contains search 'id'            = false
[DEBUG] Change matches search criteria to apply
[DEBUG] quickXorHash is missing
[DEBUG] sha256Hash is missing
ERROR: OneDrive API inconsistency - the file does not have any hash
[DEBUG] Flagging as unwanted: find(item.parentId).length != 0
[DEBUG] Skipping OneDrive change as this is determined to be unwanted
```


So the entry *before* the error should give you information about what is failing...

**Comment by pablobass** [comment]
> > If you are seeing an 'id' rather than a 'path' this means the item in question is being excluded by your configuration.
> 
> When I was playing with the code I think to remember that sometimes, even if the `->parent` field was not available, the `name` field could have some information... but I was really printing random things so I can't guarantee it.
> 
> @pablobass in the "double verbose" log you should see something like this:
> 
> ```
> [DEBUG] Not a OneDrive deleted item change
> [DEBUG] defaultRootId                                        = 01CRET2R56Y2GOVW7725BZO354PWSELRRZ
> [DEBUG] 'search id'                                          = 01CRET2R56Y2GOVW7725BZO354PWSELRRZ
> [DEBUG] id == defaultRootId                                  = true
> [DEBUG] isItemRoot(item)                                     = false
> [DEBUG] item['name'].str == 'root'                           = false
> [DEBUG] singleDirectoryScope                                 = false
> [DEBUG] 'parentReference id'                                 = 01CRET2R63B3UZEGQYDFFIZCQVNQAURE4X
> [DEBUG] search criteria: syncFolderName                      = root
> [DEBUG] search criteria: syncFolderPath                      = 
> [DEBUG] search criteria: syncFolderChildPath                 = 
> [DEBUG] thisItemId                                           = 01CRET2R2NSM7N4LKUYVGKU7PU3ZBUEZ73
> [DEBUG] thisItemParentPath                                   = /drives/b!LtH***************nhNqI5o-VcAsHJ_LJBMtZ3mQagpnag-F9Cq/root:/Blocs de notas/Cuaderno de Moodle
> [DEBUG] thisItemFullPath                                     = /drives/b!LtH***************nhNqI5o-VcAsHJ_LJBMtZ3mQagpnag-F9Cq/root:/Blocs de notas/Cuaderno de Moodle/Cuaderno de Moodle.onetoc2
> [DEBUG] 'item id' matches search 'id'                        = false
> [DEBUG] 'parentReference id' matches search 'id'             = false
> [DEBUG] 'thisItemParentPath' contains 'syncFolderChildPath'  = true
> [DEBUG] 'thisItemParentPath' contains search 'id'            = false
> [DEBUG] Change matches search criteria to apply
> [DEBUG] quickXorHash is missing
> [DEBUG] sha256Hash is missing
> ERROR: OneDrive API inconsistency - the file does not have any hash
> [DEBUG] Flagging as unwanted: find(item.parentId).length != 0
> [DEBUG] Skipping OneDrive change as this is determined to be unwanted
> ```
> 
> So the entry _before_ the error should give you information about what is failing...

Debug info from one of the "hash" notices:

`[DEBUG] OneDrive Change: {"@odata.type":"#microsoft.graph.driveItem","cTag":"aYzpBRTU3NDNCQkJFMzA3RDBCITE0MjAzNi4yNTc","deleted":{"state":"hardDeleted"},"eTag":"aQUU1NzQzQkJCRTMwN0QwQiExNDIwMzYuMQ","file":{"hashes":{"sha1Hash":"A86F23B202C2223E3219D0963660EF85E0C33572"}},"id":"AE5743BBBE307D0B!142036","name":"","parentReference":{"driveId":"ae5743bbbe307d0b","driveType":"personal","id":"AE5743BBBE307D0B!142035"},"size":235644}`

and in my config:

`skip_file = "*.tmp|*.crdownload|desktop.ini|thumbs.db|.~*|.DS_Store|._.DS_Store"`

**Comment by ificator** [comment]
@pablobass the `OneDrive Change` JSON snippet suggests this file is hard deleted (i.e. unrecoverable and simply exists so that we can tell clients that it's gone). It's possible that the process of generating the `quickXorHash` values for older files simply ignored these (because it's likely the blob doesn't exist in the store anymore). Can you confirm whether this is the case for all of the files that are generating the missing hash message for you?

**Comment by pablobass** [comment]
> @pablobass the `OneDrive Change` JSON snippet suggests this file is hard deleted (i.e. unrecoverable and simply exists so that we can tell clients that it's gone). It's possible that the process of generating the `quickXorHash` values for older files simply ignored these (because it's likely the blob doesn't exist in the store anymore). Can you confirm whether this is the case for all of the files that are generating the missing hash message for you?

Like I said, I can't tell wich files the "hash" notice are referring to. Even with "double verbose" option.

**Comment by ificator** [comment]
@pablobass the double verbose is giving the `OneDrive Change` JSON blob which lets us see that file `AE5743BBBE307D0B!142036` is hard deleted. If we can confirm that's the case for all of the other item ids that would let us scope the problem down to a case where it'll resolve itself over time.

@abraunegg out of curiosity, why do we hash deleted files? Is that so you don't delete the local content if it appears to be different to the cloud? I'm trying to rationalize if we can hide the hash errors for this case, but realistically this hash check still makes sense.

**Comment by pablobass** [comment]
> @pablobass the double verbose is giving the `OneDrive Change` JSON blob which lets us see that file `AE5743BBBE307D0B!142036` is hard deleted. If we can confirm that's the case for all of the other item ids that would let us scope the problem down to a case where it'll resolve itself over time.
> 
> @abraunegg out of curiosity, why do we hash deleted files? Is that so you don't delete the local content if it appears to be different to the cloud? I'm trying to rationalize if we can hide the hash errors for this case, but realistically this hash check still makes sense.

I'm almost sure that all the other files gives me the "same" information (with a different ID, of course). I didn't post the whole log because of this, and I thought it'll be too long for a comment. Anyway, of course I can doy it if you need.

**Comment by abraunegg** [comment]
@ificator 
> @abraunegg out of curiosity, why do we hash deleted files? Is that so you don't delete the local content if it appears to be different to the cloud? I'm trying to rationalize if we can hide the hash errors for this case, but realistically this hash check still makes sense.

The 'onedrive' application does not hash deleted files.

What is happening is, in the /delta call, depending on the users configuration (where they can configure client side filtering rules to only sync content that they are interested in) - the call to /delta will not include the deltaLink URL, thus all object items are returned - including any deleted items or other data.

When the application is processing the JSON responses, it is processing each response, and creating an item struct that the application uses for that object. It is in that function call where this warning is being generated - because the JSON item as received from OneDrive does not contain either a `quickXorHash` or a `sha256hash` JSON entry item.

Essentially because the API is not providing the data in the /delta response - that is why this message is being presented.

**Comment by pablobass** [comment]
I can add one more thing to this: "Hash" notices only appear when I use "onedrive --monitor", but not when I execute "onedrive --synchronize". Weird, right?

**Comment by abraunegg** [comment]
@pablobass 

> I can add one more thing to this: "Hash" notices only appear when I use "onedrive --monitor", but not when I execute "onedrive --synchronize". Weird, right?

Please submit a verbose debug log for analysis following the correct support process.

**Comment by ificator** [comment]
> The 'onedrive' application does not hash deleted files.
> 
> What is happening is, in the /delta call, depending on the users configuration (where they can configure client side filtering rules to only sync content that they are interested in) - the call to /delta will not include the deltaLink URL, thus all object items are returned - including any deleted items or other data.
> 
> When the application is processing the JSON responses, it is processing each response, and creating an item struct that the application uses for that object. It is in that function call where this warning is being generated - because the JSON item as received from OneDrive does not contain either a `quickXorHash` or a `sha256hash` JSON entry item.
> 
> Essentially because the API is not providing the data in the /delta response - that is why this message is being presented.

@abraunegg in that case, if the item contains the `deleted` facet, could we skip the logic that asserts the required hashes are present? That'll avoid some noise in your client logs while this type of hash-less data ages out of the system.

**Comment by abraunegg** [comment]
@ificator 
> @abraunegg in that case, if the item contains the `deleted` facet, could we skip the logic that asserts the required hashes are present? That'll avoid some noise in your client logs while this type of hash-less data ages out of the system.

In the current v2.4.x codebase, unfortunately this is not going to be possible without a new v2.4.26 release.

In the v2.5.x codebase (which is in active development & is a 100% re-write of the client to implement some major features & change) this has already factored in - and only items that are files and not deleted are checked for the presence of the required hash items.

**Comment by ificator** [comment]
I'm going to close this issue as I think the primary problem has been addressed, and the updated client will handle the deleted files appropriately.

**Comment by futymax** [comment]
Hello! 
Today I was testing out the client and got this same error (not sure if it was the same source tho). After reading the log outputted with both double verbose and https was able to find that OneDrive does supply information about files in the Personal Vault in the Delta api.

Testing in a freshly downloaded copy of onedrive, single folder, download only and resync:
Part of first HTTP call `* h2h3 [:path: /v1.0/drives/xxxxxx/root:/NasSync?select=id,name,eTag,cTag,deleted,file,folder,root,fileSystemInfo,remoteItem,parentReference,size]
`
A few calls later, it reaches the offending file
`[DEBUG] OneDrive Change: {"cTag":"xxxxxx","deleted":{"state":"hardDeleted"},"eTag":"xxxxxx","file":{"hashes":{"sha1Hash":"xxxxxx"}},"id":"xxxxxx!150524","name":"","parentReference":{"driveId":"xxxxxx","driveType":"personal","id":"xxxxxx!146271"},"size":1320}`

And the corresponding warning just after
`WARNING: OneDrive API inconsistency - this file does not have any hash: xxxxxx!150524`

And there is some information about the folder 146271
`[DEBUG] OneDrive Change: {"cTag":"xxxxxx","deleted":{"state":"hardDeleted"},"eTag":"xxxxxx","folder":{"childCount":0,"view":{"sortBy":"name","sortOrder":"ascending","viewType":"thumbnails"}},"id":"xxxxxx!146271","name":"","parentReference":{"driveId":"xxxxxx","driveType":"personal","id":"xxxxxx!246"},"size":0}
[DEBUG] isRoot                                               = false
[DEBUG] !hasParentReferenceId(item)                          = false
[DEBUG] isItemDeleted(item)                                  = true
[DEBUG] Handling change as 'root item', has no parent reference or is a deleted item
[DEBUG] Remote deleted item
[DEBUG] The item we are syncing is a folder`

And that folder is directly inside the root directory (246).

After some digging in Google about the meaning of the "hardDeleted" value

**Comment by pablobass** [comment]
Sorry for reviving this old thread.

I discovered which files were returning the warning 'OneDrive API inconsistency - this file does not have any hash.'

They were not deleted files, but rather those located in the 'Personal Vault.'

Evidently, the Linux client could somehow list them but not view them or something like that. I'm not a programmer.

I’m sharing this just in case it’s helpful information, as to this day (onedrive v2.4.25-13-g1a88d33) I was still seeing those warnings.

Finally, I was able to get rid of those alerts in the logs, and I'm happy about that, but unfortunately, it was by removing the files I had in the vault.

Maybe this issue has been resolved since version 2.5.x, but I don't know since I'm waiting for the final version to update.

Anyway, I hope this can be of help.

Best regards.

**Comment by abraunegg** [comment]
@pablobass 
> Sorry for reviving this old thread.
> 
> I discovered which files were returning the warning 'OneDrive API inconsistency - this file does not have any hash.'
> 
> They were not deleted files, but rather those located in the 'Personal Vault.'
> 
> Evidently, the Linux client could somehow list them but not view them or something like that. I'm not a programmer.
> 
> I’m sharing this just in case it’s helpful information, as to this day (onedrive v2.4.25-13-g1a88d33) I was still seeing those warnings.
> 
> Finally, I was able to get rid of those alerts in the logs, and I'm happy about that, but unfortunately, it was by removing the files I had in the vault.
> 
> Maybe this issue has been resolved since version 2.5.x, but I don't know since I'm waiting for the final version to update.
> 
> Anyway, I hope this can be of help.
> 
> Best regards.

There is zero issue with v2.5.x version for this issue.

My test account as the following files in the vault:
![image](https://github.com/abraunegg/onedrive/assets/4956234/b911e65c-be3c-4415-b795-4282cfbc9e2c)

With v2.4.25 , the following occurs:
```
Configuration file successfully loaded
Configuring Global Azure AD Endpoints
The item database is incompatible, re-creating database table structures
Initializing the Synchronization Engine ...
Syncing changes and items from OneDrive ...
WARNING: OneDrive API inconsistency - this file does not have any hash: 66D53BE8A5056ECA!431414
WARNING: OneDrive API inconsistency - this file does not have any hash: 66D53BE8A5056ECA!431415
WARNING: OneDrive API inconsistency - this file does not have any hash: 66D53BE8A5056ECA!431416
Performing a database consistency and integrity check on locally stored data ... 
Uploading differences of ~/OneDrive
Uploading new items of ~/OneDrive
Sync with OneDrive is complete
```

With v2.5.x (specifically RC2 at the moment) , the following occurs:
```
Reading configuration file: /home/alex/.config/onedrive/config
Configuration file successfully loaded
Using IPv4 and IPv6 (if configured) for all network operations
Attempting to contact Microsoft OneDrive Login Service
Successfully reached Microsoft OneDrive Login Service
Configuring Global Azure AD Endpoints
The item database is incompatible, re-creating database table structures
Fetching items from the OneDrive API for Drive ID: 66d53be8a5056eca ....
Processing 513 applicable changes and items received from Microsoft OneDrive ..
Performing a database consistency and integrity check on locally stored data ..
Scanning the local file system '~/OneDrive' for new data to upload .
Performing a last examination of the most recent online data within Microsoft OneDrive to complete the reconciliation process
Fetching items from the OneDrive API for Drive ID: 66d53be8a5056eca .
No changes or items that can be applied were discovered while processing the data received from Microsoft OneDrive

Sync with Microsoft OneDrive is complete
Shutting down Database instance
Shutdown of Database instance is complete
```

So in v2.4.x, the API data is analysed, and the Personal Vault items kick out a warning that there is no hash

In v2.5.x, the API data is analysed, and items that are hard deleted or in the Vault are correctly handled without a warning message being displayed.
